git/util.py

```diff
@@ -150,6 +150,7 @@ def wrapper(self: "Remote", *args: Any, **kwargs: Any) -> T:
 
 @contextlib.contextmanager
 def cwd(new_dir: PathLike) -> Generator[PathLike, None, None]:
+    """Context manager to temporarily change directory. Not reentrant."""
```

Comment:
I thought it was interesting to share my thought on the word 'reentrant' , as to me it seems it actually is reentrant and it will work deterministically. It's not advisable to use this when running tests in parallel though merely because its effects are observable within the whole process.

Please do ignore, it doesn't matter :D.

Reply:
@Byron I know you've suggested I ignore this, but I realized there's actual an opportunity for improvement in what I wrote here. The important thing is that `cwd` cannot be used in quite all the ways the 3.11+ [`contextlib.chdir`](https://docs.python.org/3/library/contextlib.html#contextlib.chdir) can. I've included a docstring improvement in #1657.

Functions decorated with `contextlib.contextmanager` can be used in nested `with`-statements, but each call returns a non-reentrant context manager object. Entering it when it it has already been entered produces:

```text
AttributeError: '_GeneratorContextManager' object has no attribute 'args'
```

The simplest example of reentrancy `cwd` doesn't support looks like this, though no one would do this:

```python
cm = cwd("/path/to/dir")
with cm:
    with cm:
        ...
```

However, situations resembling this (but with more statements between the `with`s) could arise in practice:

```python
cm = cwd("/path/to/dir")
with cm:
    with cwd("/path/to/other/dir"):
        with cm:
            ...
```

`cwd` doesn't support that. Probably no one would assume it does, except there may be a temptation to conceptualize it as a backport of `contextlib.chdir`, which *does* support that. Anyway, this is still fine:

```python
with cwd("/path/to/dir"):
    with cwd("/path/to/other/dir"):
        with cwd("/path/to/dir"):
            ...
```

(Btw, I hope the ping is not an annoyance. I don't think GitHub otherwise notifies for new messages in *resolved* review threads, but I'm not sure.)

Reply:
No worries about the ping, I appreciate the explanation as indeed wasn't aware that for the context-manager implementation here isn't re-entrant, cannot be nested. It's mind-boggling that this limitation applies as it's not obvious. Nonetheless, it's not at all a problem here either, like you said.

Reply:
I think the Python design choice that makes this unintuitive is the existence of `__enter__`. Some context manager objects, like files returned by `open`, acquire their resource on *construction*, with a trivial `__enter__`. Others acquire their resource on `__enter__` (and of those, some are resuable and reentrant, some are reusable and non-reentrant, and some are neither reusable nor reentrant). But the problem context managers exist to solve, caused by lack of RAII, is that destruction, *unlike construction*, isn't deterministic.

I personally like better the approaches in C# (`Dispose` methods and `using` statements) and Java (the `AutoCloseable` interface and `try`-with-resources). Like Python, those languages lack RAII and provide special syntax to do something that would otherwise have to be done with explicit `try`-`finally`. But with the corresponding features in those languages, constructing the object is expected to acquire the resource (at least conceptually, and in most cases actually), and the only special semantics pertain to *releasing* them.